src/KeyRequestHandler.js

```diff
@@ -0,0 +1,132 @@
+/*
+Copyright 2017 Vector Creations Ltd
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import sdk from './index';
+import Modal from './Modal';
+
+export default class KeyRequestHandler {
+    constructor(matrixClient) {
+        this._matrixClient = matrixClient;
+
+        // the user/device for which we currently have a dialog open
+        this._currentUser = null;
+        this._currentDevice = null;
+
+        // userId -> deviceId -> [keyRequest]
+        this._pendingKeyRequests = {};
+    }
+
+    handleKeyRequest(keyRequest) {
+        const userId = keyRequest.userId;
+        const deviceId = keyRequest.deviceId;
+        const requestId = keyRequest.requestId;
+
+        if (!this._pendingKeyRequests[userId]) {
+            this._pendingKeyRequests[userId] = {};
```

Comment:
Whilst we enforce (somewhat) the syntax of user IDs (starting with `@`), AFAIK we have no restrictions whatsoever on the syntax of device IDs. This means you need to be careful when creating lookup tables like this, because `{}` is not empty:
```js
> s = {}
{}
> Boolean(s["hasOwnProperty"])
true
```

Prefer `Object.create(null)` which does not inherit from the default Object prototype and is therefore a true map:
```js
> s = Object.create(null)
{}
> Boolean(s["hasOwnProperty"])
false
```